Means to expose attributes and members of built-in types

jeff5 opened the pull request:
Another chunk of the exposure mechanism, following up from #205 and only slightly less of a leap. (#205 stopped at methods.)

@fwierzbicki : pinging you again for your interest.

jeff5 committed:
Implement Exposed.Member

We define exposure for members of built-in types by addition of the
mechanisms to TypeExposer. There is a fairly thorough test.

jeff5 committed:
Implement exposed get, set and delete

We define exposure for attributes (with implementor-defined protocols)
of built-in types by addition of the mechanisms to TypeExposer. There is
a fairly thorough test.

jeff5 committed:
Further test specification of Exposed.Member, Getter, etc.

We beef up a test of the mechanism for exposing types. Also, use the
annotations in a few places. (Many others potentially.)

fwierzbicki commented:
@jeff5 This looks great! Also (I know you've been doing this all along but) I want to say that I especially love how you put in comments like "Compare CPython descr_setcheck in descrobject.c" <- these pointers will be invaluable to future contributors.

jeff5 commented:
I think there is some more to do here to get the annotations working as desired. In particular we ought to be able to inherit down the Java line (as well as Python inheritance through the `__dict__` that _does_ work) so that we don't have to define members repeatedly. This comes up in the failure to expose fields Java-inherited from the base `Descriptor` class, because it is not a Python class.

jeff5 committed:
Ensure that Java-inherited fields can be exposed

Previously, the TypeExposer would miss Member annotations added to
fields in the super-class of an exposed type. We demonstrate success in
an improved test. A real example is the __name__ attribute of descriptor
objects which we like to inherit from Descriptor.

jeff5 committed:
Strengthen get-set descriptor exposure test

In particular, we test that get-set attribute definitions can be
Java-inherited.

jeff5 committed:
Merge branch 'main' into expose-bi-attr-3.8

Merging from main to fix Python version

jeff5 commented:
Although I usually leave these a few weeks after the final change, the bulk of this has had that chance so will merge shortly.
